docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,1186 @@
+      
+      OnUpdateData old_data = 2;
+      OnUpdateData new_data = 3;
+    }
+
+    message OnUpdateResponse {
+      google.protobuf.Any additional_parameters = 1;
+    }
+
+    message OnDeleteRequest {
+      string typeinstance_id = 1;
+      google.protobuf.Any additional_parameters = 2;
+    }
+
+    message OnDeleteResponse {}
+
+    message GetValueRequest {
+      string typeinstance_id = 1;
+      string resource_version_id = 2;
+      google.protobuf.Any additional_parameters = 3;
+    }
+
+    message GetValueResponse {
+      google.protobuf.Any value = 1;
+    }
+
+
+    // lock messages
+
+    message GetLockedByRequest {
+      string typeinstance_id = 1;
+      google.protobuf.Any additional_parameters = 2;
+    }
+
+    message GetLockedByResponse {
+      string locked_by = 1;
+    }
+
+    message OnLockUnlockRequest {
+      string typeinstance_id = 1;
+      google.protobuf.Any additional_parameters = 2;
+      string locked_by = 3;
+    }
+
+    message OnLockUnlockResponse {}
+
+    // services
+
+    service StorageBackend {
+      // value
+      rpc GetValue(GetValueRequest) returns (GetValueResponse);
+      rpc OnCreate(OnCreateRequest) returns (OnCreateResponse);
+      rpc OnUpdate(OnUpdateRequest) returns (OnUpdateResponse);
+      rpc OnDelete(OnDeleteRequest) returns (OnDeleteResponse);
+
+      // lock
+      rpc GetLockedBy(GetLockedByRequest) returns (GetLockedByResponse);
+      rpc OnLock(OnLockUnlockRequest) returns (OnLockUnlockResponse);
+      rpc OnUnlock(OnLockUnlockRequest) returns (OnLockUnlockResponse);
+    } 
+    ```
+
+    </details>
+
+    An implementation of such service may vary between two use cases:
+
+    1. CRUD operations on output TypeInstance actually manages external resource (e.g. Vault) -> onCreate, onUpdate, and onDelete actually creates, updates and deletes a given resource.
+    1. output TypeInstance represents external resources managed in different way (e.g. via Capact actions - like Helm Runner). IMO we shouldn't move actual Helm release installation to TypeInstance "constructor").
+
+        - The service can also implement watch for external resources (e.g. Kubernetes secrets) and call `createTypeInstances` and `deleteTypeInstances` Hub mutations. We may provide Go framework to speed up such development, similarly as we have with Runner concept.
+
+1. The service could be implemented using one of the following solutions, or other alternatives:
+
+  - [Dapr secrets](https://docs.dapr.io/developing-applications/building-blocks/secrets/secrets-overview/)
+  - [Kubernetes external secrets](https://github.com/external-secrets/kubernetes-external-secrets)
+  - [vault-k8s](https://github.com/hashicorp/vault-k8s)
+  - [db](https://upper.io/v4/getting-started/)
+  - [go-cloud](https://github.com/google/go-cloud)
+  - [stow](https://github.com/graymeta/stow)
+
+## Configuring default storage backends
+
+1. System Administrator configures default backends in Policy.
```

Comment:
Multiple - as it is not only about default static TI storage backend, but also dynamic (`common.inject.requiredTypeInstances`).

But I will rephrase it to
```
System Administrator configures default backend in Policy for static TypeInstance values.
```

Reply:
hm.. so IMO for reader it could be misleading as you use term **default backend** in previous section to highlight the **default** property in Policy. But here you refer to ** common** property in Policy. Please rephrase it once again to sth related directly with **common** 👍

Reply:
I rephrased it to:

```
## Configuring storage backends

The storage backends configuration consists of two different parts: default backend storage for TypeInstances with static values, and common TypeInstance injection.

1. System Administrator configures default backend in Policy for TypeInstances with static values.
 (...)

1. To make it easy to define injection of backend storages for TypeInstances with dynamic values, in Global and Action Policies we introduce additional feature - common TypeInstance injection configuration:
```